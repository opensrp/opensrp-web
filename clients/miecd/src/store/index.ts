import { connectReducer, getConnectedStore } from '@onaio/connected-reducer-registry';
import { gateKeeperReducer, gateKeeperReducerName } from '@onaio/gatekeeper';
import reducerRegistry, { combine, Registry } from '@onaio/redux-reducer-registry';
import session, { reducerName as sessionReducer } from '@onaio/session-reducer';

/** Initial reducers in the reducer registry */
const defaultReducers: Registry = {
    router: connectReducer as any /** Dirty hack because connectRouter LocationChangeAction does not extend Redux.AnyAction */,
};

/** Add users reducer to registry */
defaultReducers[sessionReducer] = session;
defaultReducers[gateKeeperReducerName] = gateKeeperReducer;

/** The initial store for the reveal web app */
const store = getConnectedStore(defaultReducers);

/** Set listener to add reducers to store when registered */
reducerRegistry.setChangeListener(reducers => {
    store.replaceReducer(combine(reducers));
});

export default store;
