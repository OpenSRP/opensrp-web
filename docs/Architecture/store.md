# overview

This documentation will take you through the process of creating and working with:

    - reducers
    - actions
    - selectors

## Where do I find the store

the store reducer modules are found within src/store. The initial directory structure looks something like this: (Note: this structure only depicts the initial directory structure and will as expected contain more directories and files as more are added with time)

```
...
├── src
│   ├── ...
│   ├── store
│   │   ├── ducks
│   │   │    ├── superset.ts
│   │   │    ├── tests/
│   │   ├── tests
│   │   ├── index.ts
│   │   └── selectors.ts
...
```

A single reducer module will contain the afore mentioned parts i.e. its respective actions, action creators if any, its selectors and the reducer function

## Actions

An action can be any object with a type property, they are the only way provisioned for changing the state in a redux store.

### Adding Actions

Its a common convention as well as recommended one to Use type constants, [read more on this here]().

steps:

- Add An interface describing the action structure: redux exposes an `AnyAction` type that can be extended to write custom action interfaces
- Create a general actiontypes to be used to type check all actions that will be passed to the reducer function as an argument. this will be in the form of [intersection type]()

## Selectors

whereas the action puts data in the store; the selectors on the other hand act as getters of the slice of data that this reducer module handles

selectors are functions that will take at least one argument, the store, Then based on any other arguments like say an id, can be used to filter down the data retrieved.

## The Reducer

The reducer function, takes 2 arguments; the state and an action.
using the type property of the action, the reducer should then know how to modify the state.
such functionality could be implemented as a form of switch strcture such as a list of if statements or the more recommended approach which is to use the actual switch statement

### Adding the Reducer

- Define an interface for the data to be stored in the store
- Define an interface for the store state.
- write the actual reducer function, while applying the required type annotations as required.
