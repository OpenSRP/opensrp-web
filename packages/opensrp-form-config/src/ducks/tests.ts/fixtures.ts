export const fixManifestReleases = [
    {
        identifier: '1.0.11',
        json: '{"forms_version":"1.0.11","identifiers":["blood_screening.json"}',
        appId: 'org.smartregister.reveal',
        appVersion: '1.2.14',
        createdAt: '2020-06-08T15:00:38.413+02:00',
        updatedAt: '2020-06-08T15:00:38.413+02:00',
    },
    {
        identifier: '1.0.12',
        json: '{"forms_version":"1.0.12","identifiers":["blood_screening.json"}',
        appId: 'org.smartregister.reveal',
        appVersion: '1.2.14',
        createdAt: '2020-06-08T15:03:00.950+02:00',
        updatedAt: '2020-06-08T15:03:00.950+02:00',
    },
    {
        identifier: '1.0.16',
        json: '{"forms_version":"1.0.16","identifiers":["blood_screening.json"}',
        appId: 'org.smartregister.reveal',
        appVersion: '1.2.16',
        createdAt: '2020-06-09T14:40:04.189+02:00',
        updatedAt: '2020-06-09T14:40:04.189+02:00',
    },
];
