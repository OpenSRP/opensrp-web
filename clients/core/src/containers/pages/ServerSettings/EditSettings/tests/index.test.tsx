import reducerRegistry from '@onaio/redux-reducer-registry';
import {
    locationReducerName,
    locationsReducer,
    settingsReducer,
    settingsReducerName,
    fetchLocSettings,
    fetchLocs,
} from '@opensrp/population-characteristics';
import { OpenSRPService } from '@opensrp/server-service';
import { mount, shallow } from 'enzyme';
import flushPromises from 'flush-promises';
import { createBrowserHistory } from 'history';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { Helmet } from 'react-helmet';
import { Provider } from 'react-redux';
import { Router } from 'react-router';
import { EditServerSettings } from '..';
import store from '../../../../../store';
import { allSettings, locHierarchy } from './fixtures';
import toJson from 'enzyme-to-json';

jest.mock('../../../../../configs/env');

const history = createBrowserHistory();

/** register the reducers */
reducerRegistry.register(settingsReducerName, settingsReducer);
reducerRegistry.register(locationReducerName, locationsReducer);

describe('containers/pages/ConfigForm/manifest/draftFiles', () => {
    beforeEach(() => {
        jest.resetAllMocks();
    });

    it('renders without crashing', () => {
        shallow(<EditServerSettings />);
    });

    it('renders table correctly', () => {
        /* eslint-disable-next-line @typescript-eslint/no-var-requires */
        const envModule = require('../../../../../configs/env');
        envModule.OPENSRP_API_BASE_URL = 'https://anc-stage.smartregister.org/opensrp/rest/';
        envModule.OPENSRP_API_V2_BASE_URL = 'https://anc-stage.smartregister.org/opensrp/rest/v2/';

        const wrapper = mount(
            <Provider store={store}>
                <Router history={history}>
                    <EditServerSettings />
                </Router>
            </Provider>,
        );

        const helmet = Helmet.peek();
        expect(helmet.title).toEqual('Server Settings');
        expect(wrapper.find('Connect(EditSetings)').props()).toMatchSnapshot();
    });
});
