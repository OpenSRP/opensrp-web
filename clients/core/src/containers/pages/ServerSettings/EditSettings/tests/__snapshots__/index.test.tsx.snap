// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/pages/ConfigForm/manifest/draftFiles renders table correctly 1`] = `
Object {
  "LoadingComponent": <Ripple
    borderColor="#ff5d00"
    borderStyle="solid"
    borderWidth="4px"
    height="64px"
    minHeight="80vh"
    width="64px"
  />,
  "baseURL": "https://anc-stage.smartregister.org/opensrp/",
  "getPayload": [Function],
  "locationsEndpoint": "location/location-tree",
  "restBaseURL": "https://anc-stage.smartregister.org/opensrp/rest/",
  "secAuthEndpoint": "security/authenticate",
  "settingsEndpoint": "settings/",
  "v2BaseUrl": "https://anc-stage.smartregister.org/opensrp/rest/v2/",
}
`;
